pkg/controller/usersignup/usersignup_controller.go

```diff
@@ -193,6 +204,37 @@ func (r *ReconcileUserSignup) Reconcile(request reconcile.Request) (reconcile.Re
 	return reconcile.Result{}, nil
 }
 
+func (r *ReconcileUserSignup) generateCompliantUsername(instance *toolchainv1alpha1.UserSignup) (string, error) {
+	replaced := strings.ReplaceAll(strings.ReplaceAll(instance.Spec.Username, "@", "-at-"), ".", "-")
+
+	errs := validation.IsQualifiedName(replaced)
+	if len(errs) > 0 {
+		return "", NewSignupError(fmt.Sprintf("transformed username [%s] is invalid", replaced))
+	}
+
+	transformed := replaced
+
+	for i := 1; i < 101; i++ { // No more than 100 attempts to find a vacant name
+		mur := &toolchainv1alpha1.MasterUserRecord{}
+		// Check if a MasterUserRecord exists with the same transformed name
+		namespacedMurName := types.NamespacedName{Namespace: instance.Namespace, Name: transformed}
+		err := r.client.Get(context.TODO(), namespacedMurName, mur)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				return "", err
+			}
+			// If there was a NotFound error looking up the mur, it means we found an available name
+			return transformed, nil
+		} else if mur.Spec.UserID == instance.Name { // If the found MUR has the same UserID as the UserSignup, then *it* is the correct MUR
```

Comment:
1. We use label now to store UserID.
2. If for whatever reason there is already MUR with the same UserID label then we should return an error here. Theoretically it can happen if there are multiple reconciles executed in the same time for the same UserSignup resource. But in this case returning an error (which will re-queue) will work. The next reconcile will be successful.

Reply:
Fixed.